### PR TITLE
Created PhysicalTicket model

### PR DESCRIPTION
**Pull request by siddhantbajaj**
This model represents a ticket of conference. So now we have three models to handle tickets:

1. Ticket: It represent different types of ticket available for a conference.
2. TicketPuchase: It represent all the Physical tickets bought in single purchase by a user.
3. PhysicalTicket: It will represent a single purchased ticket of a conference.

### CODE REVIEW

File: app/helpers/physical_ticket_helper.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+module PhysicalTicketHelper
```
Comment (Ana06) [inline app/helpers/physical_ticket_helper.rb]: I guess there are aome classes that were automatically generated, but if by now they are not use, like it is the case with this helper, I would better remove them 🤔

File: app/models/physical_ticket.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class PhysicalTicket < ActiveRecord::Base
+	belongs_to :conference
+	belongs_to :user
+	belongs_to :ticket_purchases
+	belongs_to :ticket
```
Comment (Ana06) [inline app/models/physical_ticket.rb]: A ticket purchase already belongs to ticket, right? So you don't need this relation with ticket. In case that at some point the ticket is needed we can introduce a thought ticket_purchase relation that do not need extra fields in the database. :wink:

File: app/models/ticket.rb [hunk 3]
```diff
@@ -1,6 +1,7 @@
 class Ticket < ActiveRecord::Base
   belongs_to :conference
   has_many :ticket_purchases, dependent: :destroy
+  has_many :physical_tickets, dependent: :destroy
```
Comment (Ana06) [inline app/models/ticket.rb]: Same here, this is not really needed. :wink:

File: app/assets/javascripts/physical_ticket.js [hunk 4]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (Ana06) [inline app/assets/javascripts/physical_ticket.js]: This file is not needed

File: app/assets/stylesheets/physical_ticket.css.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PhysicalTicket controller here.
```
Comment (Ana06) [inline app/assets/stylesheets/physical_ticket.css.scss]: This file is not needed

File: app/controllers/physical_ticket_controller.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class PhysicalTicketController < ApplicationController
+  before_filter :authenticate_user!, except: [:new, :create]
```
Comment (Ana06) [inline app/controllers/physical_ticket_controller.rb]: There is only the `new` action for this controller 😕

File: app/models/ability.rb [hunk 7]
```diff
@@ -86,6 +86,7 @@ def signed_in(user)
     can :index, Ticket
     can :manage, TicketPurchase, user_id: user.id
     can [:new, :create], Payment, user_id: user.id
+    can [:new, :create], PhysicalTicket, user_id: user.id
```
Comment (Ana06) [inline app/models/ability.rb]: There is not :create: action in the controller 😕

File: app/models/physical_ticket.rb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+class PhysicalTicket < ActiveRecord::Base
+	belongs_to :conference
```
Comment (Ana06) [inline app/models/physical_ticket.rb]: A ticket purchase already belongs to a conference so this is also not needed.

File: app/controllers/physical_ticket_controller.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+  load_resource :conference, find_by: :short_title
+  load_and_authorize_resource
+  authorize_resource :conference_registrations, class: Registration
+
+  def new
+  	@unassigned_ticket_purchases = current_user.ticket_purchases.unassigned.by_conference(@conference)
```
Comment (Ana06) [inline app/controllers/physical_ticket_controller.rb]: I am not sure what is the idea of this. Create all the available icket in the database as unassigned?  : thinking:

File: app/models/ticket_purchase.rb [hunk 10]
```diff
@@ -13,8 +13,12 @@ class TicketPurchase < ActiveRecord::Base
   delegate :price_cents, to: :ticket
   delegate :price_currency, to: :ticket
 
+  has_many :physical_tickets
+
   scope :paid, -> { where(paid: true) }
   scope :unpaid, -> { where(paid: false) }
+  scope :assigned, -> { where(assigned: true) }
```
Comment (Ana06) [inline app/models/ticket_purchase.rb]: The assigbed scope is not currently used, right? So I would prefer introducing it when we need it, it as some point this happens

File: test/controllers/physical_ticket_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class PhysicalTicketControllerTest < ActionController::TestCase
```
Comment (Ana06) [inline test/controllers/physical_ticket_controller_test.rb]: We are not using rails tests, but RSpec instead!

File: app/helpers/physical_ticket_helper.rb (hunk 1 above)
Comment (siddhantbajaj) [inline app/helpers/physical_ticket_helper.rb]: okay. :+1:

File: app/models/physical_ticket.rb (hunk 8 above)
Comment (siddhantbajaj) [inline app/models/physical_ticket.rb]: I'll remove it. :+1:

File: app/models/physical_ticket.rb (hunk 2 above)
Comment (siddhantbajaj) [inline app/models/physical_ticket.rb]: :+1:

File: app/models/ticket.rb (hunk 3 above)
Comment (siddhantbajaj) [inline app/models/ticket.rb]: :+1:

File: app/assets/javascripts/physical_ticket.js (hunk 4 above)
Comment (siddhantbajaj) [inline app/assets/javascripts/physical_ticket.js]: Sorry from now I'll make sure I don't commit unused files. :slightly_smiling_face:

File: app/assets/stylesheets/physical_ticket.css.scss (hunk 5 above)
Comment (siddhantbajaj) [inline app/assets/stylesheets/physical_ticket.css.scss]: :+1:

File: app/controllers/physical_ticket_controller.rb (hunk 9 above)
Comment (siddhantbajaj) [inline app/controllers/physical_ticket_controller.rb]: It was something related to assignment of users to tickets. But I'll remove it now.

File: app/models/ticket_purchase.rb (hunk 10 above)
Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: :+1:

**Comment by siddhantbajaj** [comment]
Workflow of Ticket Purchase feature:

1. Number of tickets required

screenshot from 2017-06-07 03 54 53

2. Generate Tickets

screenshot from 2017-06-07 03 55 13

3. Generate PDF/show

screenshot from 2017-06-07 03 55 36

File: app/models/physical_ticket.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
+class PhysicalTicket < ActiveRecord::Base
+	belongs_to :user
```
Comment (Ana06) [inline app/models/physical_ticket.rb]: this relation should be through `ticket_purchase`

File: app/controllers/physical_ticket_controller.rb [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    @current_ticket_purchases = current_user.current_ticket_purchases(@conference)
+  end
+
+  def show
+  end
+
+  def create
+    @ticket_purchases = current_user.ticket_purchases.by_conference(@conference)
+    @ticket_purchases.each do |ticket_purchase|
+      if !ticket_purchase.contains_physical_tickets?
```
Comment (Ana06) [inline app/controllers/physical_ticket_controller.rb]: `if !` == `unless` and it is more readable :wink:

Comment (Ana06) [inline app/controllers/physical_ticket_controller.rb]: and why not creating only the ones of the last `ticket_purchase`? 

I mean, this should be done in the ticket_purchase model. When quantity is set to 3, you should ensure that there are 3 physical tickets. :wink:

File: app/controllers/physical_ticket_controller.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
+class PhysicalTicketController < ApplicationController
```
Comment (Ana06) [inline app/controllers/physical_ticket_controller.rb]: I doubth that you need in this controller something else except index and show.

File: app/models/physical_ticket.rb [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+	belongs_to :ticket_purchase
+
+	scope :by_user, ->(user) { where(user_id: user.id) }
+
+	def ticket_type
+		ticket_purchase.ticket
+	end
+
+	def conference
```
Comment (Ana06) [inline app/models/physical_ticket.rb]: this can be done with a through association, but I would say it is not needed yet.

File: app/models/physical_ticket.rb [hunk 16]
```diff
@@ -0,0 +1,14 @@
+class PhysicalTicket < ActiveRecord::Base
+	belongs_to :user
+	belongs_to :ticket_purchase
+
+	scope :by_user, ->(user) { where(user_id: user.id) }
+
+	def ticket_type
```
Comment (Ana06) [inline app/models/physical_ticket.rb]: this can be done with a through association, but I would say it is not needed yet.

File: app/models/physical_ticket.rb [hunk 17]
```diff
@@ -0,0 +1,14 @@
+class PhysicalTicket < ActiveRecord::Base
+	belongs_to :user
+	belongs_to :ticket_purchase
+
+	scope :by_user, ->(user) { where(user_id: user.id) }
```
Comment (Ana06) [inline app/models/physical_ticket.rb]: you could create a through association in user for this.

File: app/models/ticket_purchase.rb [hunk 18]
```diff
@@ -58,4 +60,10 @@ def self.update_quantity(conference, quantity, ticket, user)
     purchase.quantity = quantity if quantity > 0
     purchase
   end
+
+  ##
+  # Return true if tickets in ticket purchase are already created.
+  def contains_physical_tickets?
+    return PhysicalTicket.exists?(ticket_purchase_id: self)
```
Comment (Ana06) [inline app/models/ticket_purchase.rb]: `return` is not needed. And this can be written as:

`physical_tickets.any?`

File: app/views/physical_ticket/_tickets.html.haml [hunk 19]
```diff
@@ -0,0 +1,21 @@
+.page-header
```
Comment (Ana06) [inline app/views/physical_ticket/_tickets.html.haml]: how all this work with all conferences where physical tickets didn't exist?

File: db/migrate/20170603095900_create_physical_tickets.rb [hunk 20]
```diff
@@ -0,0 +1,10 @@
+class CreatePhysicalTickets < ActiveRecord::Migration
+  def change
+    create_table :physical_tickets do |t|
+      t.integer :user_id, null: false
```
Comment (Ana06) [inline db/migrate/20170603095900_create_physical_tickets.rb]: this is not needed

File: app/models/physical_ticket.rb (hunk 12 above)
Comment (siddhantbajaj) [inline app/models/physical_ticket.rb]: Yes I'll make the changes. :wink:

File: app/controllers/physical_ticket_controller.rb (hunk 13 above)
Comment (siddhantbajaj) [inline app/controllers/physical_ticket_controller.rb]: if we create Physical Tickets in ticket purchase controller then basically we are creating tickets before their payment is made through stripe.

File: app/controllers/physical_ticket_controller.rb (hunk 14 above)
Comment (siddhantbajaj) [inline app/controllers/physical_ticket_controller.rb]: if we create Physical Tickets in ticket purchase controller then basically we are creating tickets before their payment is made through stripe.

File: app/models/physical_ticket.rb (hunk 17 above)
Comment (siddhantbajaj) [inline app/models/physical_ticket.rb]: I'll change this. :wink:

File: app/controllers/physical_ticket_controller.rb (hunk 13 above)
Comment (siddhantbajaj) [inline app/controllers/physical_ticket_controller.rb]: Sorry my bad did you mean we should create tickets in after_create and after_update callbacks of ticket purchase model?

Comment (Ana06) [inline app/controllers/physical_ticket_controller.rb]: exactly! :wink: after the payment is done, we set the purchase to paid right? in that moment you also should create the physical tickets :slightly_smiling_face:

File: app/models/user.rb [hunk 21]
```diff
@@ -100,6 +101,10 @@ def supports? conference
     ticket_purchases.find_by(conference_id: conference.id).present?
   end
 
+  def tickets conference
+    conference.physical_tickets.select { |t| t.user == self }
```
Comment (Ana06) [inline app/models/user.rb]: `physical_tickets.where(conference: conference)`

File: app/views/physical_ticket/_tickets.html.haml (hunk 19 above)
Comment (Ana06) [inline app/views/physical_ticket/_tickets.html.haml]: ^^

**Comment by Ana06** [comment]
@siddhantbajaj 

One important detail. There is something currently "wrong" in the payment. Check this:
https://github.com/Ana06/osem/blob/7a7fcced487de914c9bd2bfc9e4e51839413aa8e/app/controllers/payments_controller.rb#L41

`update_all` skips all validations and callbacks, so your `after_update` in the TicketPurchase model doesn't work.

So, the first thing should be correcting this, because we do not want to use method that skip validations.

What I would do is to create a pay method in the TicketPurchase model and call it from `payment_controller `method. In the pay method you will update the paid attribute and create the physical tickets.

File: app/models/physical_ticket.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
+class PhysicalTicket < ActiveRecord::Base
+  belongs_to :ticket_purchase
+
+  def ticket_type
```
Comment (mdeniz) [inline app/models/physical_ticket.rb]: This should be a :through relationship

File: app/models/physical_ticket.rb [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+	ticket_purchase.ticket
+  end
+
+  def conference
+	ticket_purchase.conference
```
Comment (mdeniz) [inline app/models/physical_ticket.rb]: This should be a :through relationship

File: app/models/physical_ticket.rb [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  def user
+	ticket_purchase.user
```
Comment (mdeniz) [inline app/models/physical_ticket.rb]: This should be a :through relationship

File: app/models/user.rb [hunk 25]
```diff
@@ -100,6 +101,10 @@ def supports? conference
     ticket_purchases.find_by(conference_id: conference.id).present?
   end
 
+  def tickets conference
```
Comment (mdeniz) [inline app/models/user.rb]: Use parenthesis for params 😄

File: spec/factories/ticket_purchases.rb [hunk 26]
```diff
@@ -4,5 +4,6 @@
     conference
     ticket
     quantity 10
+    paid false
```
Comment (mdeniz) [inline spec/factories/ticket_purchases.rb]: Does it worth to have a 'sub' factory for already paid ticket_purchases??

File: spec/models/ticket_purchase_spec.rb [hunk 27]
```diff
@@ -2,118 +2,136 @@
 
 describe TicketPurchase do
 
-  describe 'validations' do
-    it 'has a valid factory' do
-      expect(build(:ticket_purchase)).to be_valid
+  # describe 'validations' do
```
Comment (mdeniz) [inline spec/models/ticket_purchase_spec.rb]: Why those commented specs?

Comment (siddhantbajaj) [inline spec/models/ticket_purchase_spec.rb]: Sorry my bad I was testing something here so I commented all other specs and forgot to un-comment them.

File: spec/factories/ticket_purchases.rb (hunk 26 above)
Comment (siddhantbajaj) [inline spec/factories/ticket_purchases.rb]: Yes I think it would be useful to have a sub factory here for paid ticket_purchases.

File: app/models/ticket_purchase.rb [hunk 28]
```diff
@@ -58,4 +63,24 @@ def self.update_quantity(conference, quantity, ticket, user)
     purchase.quantity = quantity if quantity > 0
     purchase
   end
+
+  def pay(payment)
+    self.update_attributes(paid: true, payment_id: payment.id)
```
Comment (Ana06) [inline app/models/ticket_purchase.rb]: you can write this as:

``` Ruby
update_attributes(paid: true, payment: payment)
```

`self` is not needed. Please read: https://hackhands.com/three-golden-rules-understand-self-ruby :wink: And Rubocop can also help you, as the `Style/RedundantSelf` cop is enabled. 

And as we have a relation, you can use it instead of the `id` :slightly_smiling_face:

File: app/models/physical_ticket.rb [hunk 29]
```diff
@@ -0,0 +1,6 @@
+class PhysicalTicket < ActiveRecord::Base
+  belongs_to :ticket_purchase
+  delegate :ticket, to: :ticket_purchase
```
Comment (Ana06) [inline app/models/physical_ticket.rb]: `delegate` is for expose contained objects’ public methods as your own. Ref: https://apidock.com/rails/Module/delegate

for this you should use `:through` associations. See http://guides.rubyonrails.org/association_basics.html#the-has-one-through-association

File: spec/factories/ticket_purchases.rb [hunk 30]
```diff
@@ -4,5 +4,11 @@
...
+    factory :paid_ticket_purchase do
+    	after(:build) do |ticket_purchase|
```
Comment (Ana06) [inline spec/factories/ticket_purchases.rb]: you also need to create physical tickets here

File: spec/factories/ticket_purchases.rb (hunk 26 above)
Comment (Ana06) [inline spec/factories/ticket_purchases.rb]: It is false as default in the database, so this shouldn't be needed, right? :thinking:

File: spec/models/ticket_purchase_spec.rb [hunk 31]
```diff
@@ -116,4 +116,22 @@
       expect(message.blank?).to be true
     end
   end
+
```
Comment (Ana06) [inline spec/models/ticket_purchase_spec.rb]: This is in the wrong commit!

File: db/schema.rb [hunk 32]
```diff
@@ -11,7 +11,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 20170606001722) do
+ActiveRecord::Schema.define(version: 20170603095900) do
```
Comment (Ana06) [inline db/schema.rb]: This is in the wrong commit!

File: app/models/ticket_purchase.rb [hunk 33]
```diff
@@ -13,11 +13,16 @@ class TicketPurchase < ActiveRecord::Base
...
   delegate :price_currency, to: :ticket
 
+  has_many :physical_tickets
+
   scope :paid, -> { where(paid: true) }
   scope :unpaid, -> { where(paid: false) }
   scope :by_conference, ->(conference) { where(conference_id: conference.id) }
   scope :by_user, ->(user) { where(user_id: user.id) }
 
+  after_create :create_physical_tickets
```
Comment (mdeniz) [inline app/models/ticket_purchase.rb]: This 2 methods are unneeded if you create the tickets at the pay method

File: app/models/user.rb [hunk 34]
```diff
@@ -6,6 +6,7 @@ class UserDisabled < StandardError
 
 class User < ActiveRecord::Base
   rolify
+  has_many :physical_tickets, through: :ticket_purchases, dependent: :destroy
```
Comment (mdeniz) [inline app/models/user.rb]: I guess that we are not going to use this relationship without scoping it by conference... so, better if you do something like this:

```ruby
has_many :physical_tickets, -> (conference) { where(conference: conference) }, through: :ticket_purchases, dependent: :destroy
```

If you do so, then the **tickets** method that you created (line 104 here) is not needed.

File: app/models/user.rb [hunk 35]
```diff
@@ -100,6 +101,10 @@ def supports? conference
     ticket_purchases.find_by(conference_id: conference.id).present?
   end
 
+  def tickets(conference)
```
Comment (mdeniz) [inline app/models/user.rb]: This method will be unneeded if you scope the relationship by conference

File: app/views/conferences/_conference_details.html.haml [hunk 36]
```diff
@@ -38,3 +38,5 @@
               = link_to 'Subscribe', conference_subscriptions_path(conference.short_title), method: :post, class: 'btn btn-default'
             - else
               = link_to 'Unsubscribe', conference_subscriptions_path(conference.short_title), method: :delete, class: 'btn btn-default'
+            - if !current_user.nil? && current_user.tickets(conference).any?
```
Comment (mdeniz) [inline app/views/conferences/_conference_details.html.haml]: Would be better to use **present?** here? Also if you use the scoped relationship for **physical_tickets** it will look like this:

```ruby
if current_user.present? && current_user.physical_tickets(conference).any?
```
Or just using **try**:

```ruby
if current_user.try(:physical_tickets, conference).try(:any?)
```

File: spec/factories/ticket_purchases.rb (hunk 30 above)
Comment (mdeniz) [inline spec/factories/ticket_purchases.rb]: Instead of marking the object as paid and creating physical_tickets by hand, you can use the **pay** method, so maybe you need a payment object too.

File: spec/factories/ticket_purchases.rb [hunk 37]
```diff
@@ -4,5 +4,11 @@
...
+    		ticket_purchase.paid = true
```
Comment (mdeniz) [inline spec/factories/ticket_purchases.rb]: And BTW, normally overriding just an attribute is better to put this directly on the subfactory, not in a **after(:build)**:

```ruby
factory :paid_ticket_purchase do
  paid true
end
```

File: spec/models/ticket_purchase_spec.rb (hunk 31 above)
Comment (mdeniz) [inline spec/models/ticket_purchase_spec.rb]: Those spec will be unneeded if you remove the methods that are tested

File: app/models/ticket_purchase.rb [hunk 38]
```diff
@@ -58,4 +63,24 @@ def self.update_quantity(conference, quantity, ticket, user)
...
     purchase
   end
+
+  def pay(payment)
+    self.update_attributes(paid: true, payment_id: payment.id)
+  end
+
+  protected
+
+  def create_physical_tickets
```
Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: But then what about Free tickets they does not have any payment and paid attribute is already true.:thinking:

Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: But then what about Free ticket_purchases they does not have any payment and their paid attribute is already true.

File: spec/factories/ticket_purchases.rb (hunk 30 above)
Comment (siddhantbajaj) [inline spec/factories/ticket_purchases.rb]: I think we should create tickets in after_create method because there are free ticket purchases as well that does not have any payment and if we do that we don't have to create tickets here.

File: spec/factories/ticket_purchases.rb (hunk 37 above)
Comment (siddhantbajaj) [inline spec/factories/ticket_purchases.rb]: Sure I'll make the change. :+1:

File: app/models/ticket_purchase.rb (hunk 38 above)
Comment (Ana06) [inline app/models/ticket_purchase.rb]: for example,

instead of setting paid to `true` in https://github.com/Ana06/osem/blob/7a7fcced487de914c9bd2bfc9e4e51839413aa8e/app/models/ticket_purchase.rb#L47

in the next line you can just do:

`purchase.pay if  ticket.price_cgents.zero?`

Comment (mdeniz) [inline app/models/ticket_purchase.rb]: I agree with Ana, call the **pay** method in the purchase_ticket after creating the **TicketPurchase** object if it is free. That way you will have all the logic in one place, the **pay** method.

But rather than just doing **purchas.pay** you will need to do **purchase.pay(nil)** in the case of free tickets.

Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: > That way you will have all the logic in one place, the pay method.

This makes sense, I'll make the changes. :+1:

File: app/models/user.rb (hunk 34 above)
Comment (siddhantbajaj) [inline app/models/user.rb]: I tried this but it's not working, it gives a SQL exception because physical ticket does not have conference_id column. So I am looking for a work around for this.:thinking:

File: app/models/physical_ticket.rb (hunk 29 above)
Comment (siddhantbajaj) [inline app/models/physical_ticket.rb]: Thanks a lot for this. I've made the changes. :+1:

File: app/controllers/payments_controller.rb [hunk 39]
```diff
@@ -18,7 +18,7 @@ def create
 
     if @payment.purchase && @payment.save
       update_purchased_ticket_purchases
-      redirect_to conference_conference_registration_path(@conference.short_title),
+      redirect_to conference_physical_ticket_index_path,
```
Comment (siddhantbajaj) [inline app/controllers/payments_controller.rb]: So like `update_purchased_ticket_purchases` is still excuting in the background and the user is redirected to index page? I am not sure if redirect_to works synchronously.

File: app/views/conferences/_conference_details.html.haml (hunk 36 above)
Comment (siddhantbajaj) [inline app/views/conferences/_conference_details.html.haml]: :+1:

File: app/models/user.rb (hunk 35 above)
Comment (siddhantbajaj) [inline app/models/user.rb]: :+1:

File: app/models/ticket_purchase.rb (hunk 33 above)
Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: :+1:

File: app/models/user.rb (hunk 34 above)
Comment (Ana06) [inline app/models/user.rb]: You can add a through relation for conference also in ticket :wink:

File: app/models/physical_ticket.rb [hunk 40]
```diff
@@ -0,0 +1,9 @@
+class PhysicalTicket < ActiveRecord::Base
+  belongs_to :ticket_purchase
+  has_one :ticket, through: :ticket_purchase
+  has_one :conference, through: :ticket_purchase
+  has_one :user, through: :ticket_purchase
+
+  scope :by_conference, ->(conference) { joins(:ticket_purchase).where('ticket_purchases.conference_id = ?', conference.id) }
```
Comment (mdeniz) [inline app/models/physical_ticket.rb]: Why don't you use the relationship instead of doing a join by hand?

```ruby
scope :by_conference, ->(conference) { where(conference: conference) }
```

File: app/models/ticket_purchase.rb [hunk 41]
```diff
@@ -58,4 +60,11 @@ def self.update_quantity(conference, quantity, ticket, user)
     purchase.quantity = quantity if quantity > 0
     purchase
   end
+
+  def pay(payment)
+    update_attributes(paid: true, payment_id: payment)
```
Comment (mdeniz) [inline app/models/ticket_purchase.rb]: I guess you can assign it to **payment** instead of to  **payment_id**

```ruby
 update_attributes(paid: true, payment: payment)
```

File: app/models/ticket_purchase.rb [hunk 42]
```diff
@@ -58,4 +60,11 @@ def self.update_quantity(conference, quantity, ticket, user)
...
+    quantity.times do
+      PhysicalTicket.create(ticket_purchase_id: id)
```
Comment (mdeniz) [inline app/models/ticket_purchase.rb]: For creating related objects is better to use the relationship methods it will assign the TicketPurchase object without having to explicitly write it there. 

Also I would prefer to surround this with a transaction. And BTW even better using a one line block style for the iterating part:

```ruby
PhysicalTicket.transaction do
  quantity.times { physical_tickets.create }
end
```

File: app/models/user.rb (hunk 34 above)
Comment (mdeniz) [inline app/models/user.rb]: Does this works like Ana said using the through relationship?

File: app/models/physical_ticket.rb (hunk 40 above)
Comment (siddhantbajaj) [inline app/models/physical_ticket.rb]: Yes I tried that, but its not working. Same SQL exception no such column conference_id in physical_tickets table. 🤔

Comment (mdeniz) [inline app/models/physical_ticket.rb]: Ok then 👍

File: app/models/user.rb (hunk 34 above)
Comment (siddhantbajaj) [inline app/models/user.rb]: But I already have a through relation for conference in the physical ticket and still it didn't work. :confused:

File: app/models/ticket_purchase.rb (hunk 41 above)
Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: No I think we have to specify the exact attribute name in update_attributes. I tried this and it gives an error unknown attribute payment.

File: app/models/ticket_purchase.rb (hunk 42 above)
Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: Sure I'll make the change. 🙂

File: app/models/physical_ticket.rb (hunk 40 above)
Comment (Ana06) [inline app/models/physical_ticket.rb]: That's probably the reason why there is a relation with conference in most of our models, to avoid joins. I am not sure what is better. :thinking:

Comment (mdeniz) [inline app/models/physical_ticket.rb]: Less dangerous to do the joins, and at the same time less performance. But adding a relationship directly to the conference can end in having wrong data because the duplication.

**Comment by Ana06** [comment]
@siddhantbajaj please remember to squash the commits :wink:

File: app/models/ticket_purchase.rb (hunk 41 above)
Comment (Ana06) [inline app/models/ticket_purchase.rb]: Please check it again. You should be able to use `payment` directly as @mdeniz says.

File: app/models/user.rb (hunk 34 above)
Comment (Ana06) [inline app/models/user.rb]: you need the `join`. But anyway the by_conference method is always used like:

`user.physical_tickets.by_conference`

so you can just add the `by_conference` scope here, and even remove it from the other model, as would only be used here.

File: app/views/conferences/_conference_details.html.haml [hunk 43]
```diff
@@ -38,3 +38,5 @@
               = link_to 'Subscribe', conference_subscriptions_path(conference.short_title), method: :post, class: 'btn btn-default'
             - else
               = link_to 'Unsubscribe', conference_subscriptions_path(conference.short_title), method: :delete, class: 'btn btn-default'
+            - if current_user.present? && current_user.physical_tickets.by_conference(conference).any?
```
Comment (Ana06) [inline app/views/conferences/_conference_details.html.haml]: you don't need the `present?`, just:

``` Ruby
if current_use && current_user.physical_tickets.by_conference(conference).any?
```

File: app/models/user.rb (hunk 34 above)
Comment (siddhantbajaj) [inline app/models/user.rb]: > so you can just add the by_conference scope here, and even remove it from the other model, as would only be used here. 

Sorry I didn't understand what you mean here. we already have by_conference scope in PhysicalTicket model why do we have to remove it from there and it in this user model. 😕

File: app/models/ticket_purchase.rb (hunk 41 above)
Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: I found the mistake actually one to many association between ticket purchase and payment wasn't set up. There was no `belongs_to:payment` in ticket purchase model.

File: app/models/ticket_purchase.rb [hunk 44]
```diff
@@ -60,6 +63,13 @@ def self.update_quantity(conference, quantity, ticket, user)
     purchase.quantity = quantity if quantity > 0
     purchase
   end
+
+  def pay(payment)
+    update_attributes(paid: true, payment: payment)
+    PhysicalTicket.transaction do
```
Comment (Ana06) [inline app/models/ticket_purchase.rb]: Why do we need to wrap this into a transaction? What the difference between creating one of them or any of them?

I am not pretty sure how we should handle this. @mdeniz opinions/ideas?

File: app/models/ticket_purchase.rb [hunk 45]
```diff
@@ -2,6 +2,7 @@ class TicketPurchase < ActiveRecord::Base
   belongs_to :ticket
   belongs_to :user
   belongs_to :conference
+  belongs_to :payment
```
Comment (Ana06) [inline app/models/ticket_purchase.rb]: @siddhantbajaj why did you need to add  this?

File: app/models/ticket_purchase.rb (hunk 44 above)
Comment (Ana06) [inline app/models/ticket_purchase.rb]: @siddhantbajaj I spoke with @mdeniz and it is okey like that :wink:

File: app/models/ticket_purchase.rb (hunk 45 above)
Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: > Please check it again. You should be able to use payment directly as @mdeniz says.

It was not working because one to many association between Payment and Ticket Purchase wasn't properly set up.

Comment (Ana06) [inline app/models/ticket_purchase.rb]: Then add this is in a different commit, as it to solve something that was already wrong and not really related with this feature please. And after that I would say we are ready to merge it. @mdeniz do you want to take a last look? :wink:

Comment (siddhantbajaj) [inline app/models/ticket_purchase.rb]: Okay 🙂

File: app/views/physical_ticket/index.html.haml [hunk 46]
```diff
@@ -0,0 +1,32 @@
...
+                %td= physical_ticket.id
+                %td= physical_ticket.ticket.title
+                %td= physical_ticket.user.name
+                %td
+                  .btn-group
+                    = link_to 'Show',
+                    conference_physical_ticket_path(@conference.short_title, physical_ticket.id),
+                    class: 'btn btn-primary'
+                    = link_to 'Generate PDF',
+                    conference_physical_ticket_path(@conference.short_title, physical_ticket.id,format: :pdf),
```
Comment (Ana06) [inline app/views/physical_ticket/index.html.haml]: you could write this as

```
conference_physical_ticket_path(@conference.short_title, 
                                physical_ticket.id,format: :pdf),
```

to avoid it too be that long :wink:

**Comment by Ana06** [comment]
Merged! :tada: Great job! :+1:
